models/note.ts

```diff
@@ -0,0 +1,75 @@
+import { Schema, model } from "mongoose";
+import { ResourcesAndLinks } from "~/types";
+
+export interface INote extends Document {
+  title: string;
+  type: string; //do the same thing as the enum
+  tags: Schema.Types.ObjectId[];
+  description: string;
+  createdAt: Date;
+  creator: Schema.Types.ObjectId;
+  stars: number;
+  views: number;
+  code?: string;
+  learnings?: string[];
+  stepsToFollow?: string[];
+  content: string;
+  resourcesAndLinks: ResourcesAndLinks[];
+  relatedNotes: Schema.Types.ObjectId[];
+}
+
+//declare resources and links sub-schema
+const ResourcesAndLinksSchema = new Schema<ResourcesAndLinks>({
+  resource: { type: String, required: true },
+  url: { type: String, required: true },
+});
+
+const NoteSchema: Schema = new Schema({
+  title: {
+    type: String,
+    required: [true, "Title is required"],
+  },
+
+  type: {
+    type: String,
+    required: [true, "Type is required"],
+  },
+
+  tags: [{ type: Schema.Types.ObjectId, ref: "Tag" }],
+
+  description: String,
+
+  createdAt: {
+    type: Date,
+    required: true,
+  },
+  creator: {
+    type: Schema.Types.ObjectId,
+    ref: "User",
+    required: [true, "Creator is required."],
+  },
+  stars: Number,
+  views: Number,
+  code: {
+    type: String,
+    required: false,
+  },
+  learnings: {
+    type: [String],
+    required: false,
+  },
+  stepsToFollow: {
+    type: [String],
+    required: false,
+  },
+  content: {
+    type: String,
+    required: true,
+  },
+  resourcesAndLinks: [ResourcesAndLinksSchema],
+  relatedNotes: [{ type: Schema.Types.ObjectId, ref: "Note" }],
+});
+
+const Note = model<INote>("Note", NoteSchema);
```

Comment:
If we simply return a new model- we'll likely run into this error:

` ⨯ OverwriteModelError: Cannot overwrite `Note` model once compiled.`

The second time the model is loaded. This is because every time we try and reference this model, we'll attempt to create it again.

This is the same idea behind you using an 'isConnected' variable in database.ts

I would suggest instead importing 'models' from mongoose, so that we have access to the currently generated models- and then returning the note model if it already exists.



```suggestion
import { Schema, model, models } from "mongoose";
//....etc...
const Note = models?.Note || model<INote>("Note", NoteSchema);
```